### ISSUE TITLE
Possible Memory Leak?

### ISSUE THREAD

**Issue by kevinhammond** [input-output-hk/jormungandr#915]
**Describe the bug**
Operating system killed the jormungandr process due to an out-of-memory error after running overnight (on a Raspberry Pi Zero).

**Mandatory Information**
$ jcli --full-version
jcli 0.5.6 (HEAD-7ab929e+, release, linux [arm]) - [rustc 1.38.0 (625451e37 2019-09-23)]
$ jormungandr --full-version
jormungandr 0.5.6 (HEAD-7ab929e+, release, linux [arm]) - [rustc 1.38.0 (625451e37 2019-09-23)]

**Additional context**

Raspberry Pi Zero Rev 1.3, 512MB RAM.
 Linux version 4.4.21+ (dc4@dc4-XPS13-9333) (gcc version 4.9.3 (crosstool-NG crossto
ol-ng-1.22.0-88-g8460611) ) _redacted_ Thu Sep 15 14:17:52 BST 2016
(Raspbian Jessie: I've not got round to updating this yet - this is a spare Pi that I dug out :))

 
[jm-br1.txt](https://github.com/input-output-hk/jormungandr/files/3708042/jm-br1.txt)

**Comment by lednersin** [comment]
jormungandr using up all available memory on the linux x86_64 rustc version too. Memory jumped up along with the block info messages, then peer connection error.
Oct 09 21:37:12.000 INFO update current branch tip, date: 229.22567, parent: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx, hash: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx, task: block
...
Oct 09 21:37:24.000 INFO error connecting to peer, reason: Too many open files (os error 24), peer_addr: x.x.x.x:xxx, task: network

**Comment by mzabaluev** [comment]
@lednersin What is the jormungandr version you were using? What is the limit on the number of file descriptor (like `ulimit -n`) for the process?

**Comment by mzabaluev** [comment]
One thing to experiment with quickly is to use a recently added configuration setting `p2p.max_connections` to limit the number of open peer connections. If you still see file descriptors hogged far beyond the limit (check on Linux with `lsof -i -n -c jormungandr`), then we didn't do our homework correctly and I must investigate.

The above does not exclude the possibility of a memory leak, though we don't see this issue much noticeably yet on less memory-constrained platforms.

**Comment by kevinhammond** [comment]
Will try. As I said to @nicolas, this is a stress test system (I have larger ones) - I was just using all my spare systems as test nodes. I won’t be heartbroken if it is just too small to use in practice!  One issue on the Pi is that virtual memory is horrible since it will normally use an SD Card so you really have to keep everything in physical memory... My Pi 3 seems to have stabilised at 476M RSS, 543M Virt
(was up to 494M RSS) but it has 1GB available.  It ran for a couple of days but not seen any blocks yet.
I don’t think that is a port reuse issue as in #860, since I have generally been starting up on new port numbers.

**Comment by lednersinn** [comment]
@mzabaluev the latest 0.5.6 version, the file limit was 1024 at that time (the default soft limit for Debian, Fedora and many derivatives).
However I tried it with an with an increased 16k file limit. The block error I have wrote may have been just correlation, not causation. With the 16k limit jormungandr still jumped up to max. during the last day.

If you want to correlate with any network activity, i noticed three jumps at:
from 540 MB to 670 MB at 2019-10-10 between 10:30Z-11:00Z
from 670 MB to-870 MB at 2019-10-10 between 14:30Z-15:00Z
variations.between 870-960 MB
from 960 MB to 2 GB (max I allowed for the process) at 2019-10-11 between 06:00Z-07:00Z

(sorry, couldn't reply with the shorter name)

**Comment by jbax** [comment]
I've got the same issue (as reported [here](https://github.com/input-output-hk/shelley-testnet/issues/111)) but for me it slowly uses more and more memory (a few mb every hour), then suddenly jumps from like 350mb to 2gb in a matter of minutes. I've been running the process with the log level set to debug, hopefully you guys can use it help finding out the cause of the problem. I've uploaded the logs to my google drive [here](https://drive.google.com/open?id=1d9YFFqwopDRwDo0SWZkKaRo3VLsRqWNg), hope it can help.

Cheers!
